### ISSUE TITLE
Rest Datasource Never hits the Cache

### ISSUE THREAD

**Issue by no1melman** [apollographql/apollo-server#3935]
So according to the lines below, once a fetch is made it is stored permanently into this memoization object, therefore the HttpCache is never hit again.... 

This is bad, because we're not using all the cache logic in the HttpCache and respecting cache headers from requests etc... in fact this piece of code makes that completely and utterly redundant.

https://github.com/apollographql/apollo-server/blob/bbae424d1faec3dad25570234d672491ba0ee032/packages/apollo-datasource-rest/src/RESTDataSource.ts#L268-L278

One assumes that we should just remove it - I'm happy to make the change - just want to confirm my findings are right

**Comment by no1melman** [comment]
I've read this [blog](https://khalilstemmler.com/blogs/graphql/how-apollo-rest-data-source-caches-api-calls/)

It points out that the memoization is for request deduping. Which is fine, but one would have through that if the promise had resolved (and therefore in the cache) that you would remove the memoize and just hit the cache. That would stop multiple calls being made to the backend if one had already been made...

something like

```javascript
if (promise && !promise.IsComplete) return promise;
```

Obviously there is no `IsComplete` property, so that would need looking into

**Comment by soerenuhrbach** [comment]
**TL;DR** Make sure your datasource is constructed for each graphql request and not a singleton

---

I struggled with memoization too and thought it behaves not correctly but it does.

My project is based on [graphql-modules](https://graphql-modules.com) thats why I registered my datasource as a [provider](https://graphql-modules.com/docs/introduction/implement-providers) (which is a singleton by default).

```typescript
const PlayerModule = new GraphQLModule({
  providers: [PlayerAPI],
  typeDefs,
  resolvers,
  context: context => context
});
```

In https://github.com/apollographql/apollo-server/issues/1511#issuecomment-412523938 I found the cause of the misbehaviour:
The datasources are designed to be created for each request but here the datasource is constructed once which causes that the `memoizedResults`-map stores the responses permanently.

So I configured my datasource that it is created for every request.

```typescript
@Injectable({
  scope: ProviderScope.Request
})
export class PlayerAPI extends RESTDataSource {
...
```

After that change the memoization behaves correctly.

---

I hope I could help.
This issue can be closed since the problem is caused of incorrect use.
